src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesProducer.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec.KNN80Codec;
+
+import lombok.NonNull;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.store.FilterDirectory;
+import org.opensearch.common.io.PathUtils;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.index.engine.KNNEngine;
+import org.opensearch.knn.index.memory.NativeMemoryCacheManager;
+import org.opensearch.knn.indices.ModelCache;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
+import static org.opensearch.knn.index.mapper.KNNVectorFieldMapper.KNN_FIELD;
+
+@Log4j2
+public class KNN80DocValuesProducer extends DocValuesProducer {
+
+    private final SegmentReadState state;
+    private final DocValuesProducer delegate;
+    private final NativeMemoryCacheManager nativeMemoryCacheManager;
+    private final Map<String, String> indexPathMap = new HashMap();
+
+    public KNN80DocValuesProducer(DocValuesProducer delegate, SegmentReadState state) {
+        this.delegate = delegate;
+        this.state = state;
+        this.nativeMemoryCacheManager = NativeMemoryCacheManager.getInstance();
+
+        Directory directory = state.directory;
+        // directory would be CompoundDirectory, we need get directory firstly and then unwrap
+        if (state.directory instanceof KNN80CompoundDirectory) {
+            directory = ((KNN80CompoundDirectory) state.directory).getDir();
+        }
+        String directoryPath = ((FSDirectory) FilterDirectory.unwrap(directory)).getDirectory().toString();
```

Comment:
will this be backward compaitble?

Reply:
i think it has been backward compatible as CI tests passed

Reply:
@jmazanec15 at https://github.com/opensearch-project/k-NN/pull/1946/commits/2bcc13946a5c26a25f0183321ecdba876dc1d40a how about added try catch `ClassCastException`